daffodil-test/src/test/scala/org/apache/daffodil/layers/CheckDigit.scala

```diff
@@ -0,0 +1,205 @@
+ */
+
+package org.apache.daffodil.layers
+
+import org.apache.commons.io.IOUtils
+import org.apache.daffodil.schema.annotation.props.gen.LayerLengthKind
+import org.apache.daffodil.schema.annotation.props.gen.LayerLengthUnits
+import org.apache.daffodil.util.Maybe
+import org.apache.daffodil.processors.LayerLengthEv
+import org.apache.daffodil.processors.LayerBoundaryMarkEv
+import org.apache.daffodil.processors.LayerCharsetEv
+import org.apache.daffodil.processors.parsers.PState
+import org.apache.daffodil.processors.unparsers.UState
+
+import java.io._
+import org.apache.daffodil.dpath.NodeInfo.PrimType
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.io.DataInputStream
+import org.apache.daffodil.processors.CharsetEvBase
+import org.apache.daffodil.processors.SequenceRuntimeData
+import org.apache.daffodil.processors.SuspendableOperation
+import org.apache.daffodil.processors.charset.BitsCharsetJava
+import org.apache.daffodil.util.ByteBufferOutputStream
+import org.apache.daffodil.xml.NS
+import org.apache.daffodil.xml.RefQName
+
+import java.nio.ByteBuffer
+import java.nio.charset.Charset
+
+
+final class CheckDigitExplicit(
+  srd: SequenceRuntimeData,
+  charset: Charset,
+  layerLengthEv: LayerLengthEv)
+extends LayerTransformer() {
+  Assert.invariant(charset.newDecoder().maxCharsPerByte() == 1) // is a SBCS charset
+
+  private val checkDigitVRD = {
+    val varNamespace = NS("urn:org.apache.daffodil.layers.checkDigit")
+    val varQName = RefQName(Some("cd"), "checkDigit", varNamespace).toGlobalQName
+    val vrd = srd.variableMap.getVariableRuntimeData(varQName).getOrElse {
+      srd.SDE("Variable '%s' is not defined.", varQName.toExtendedSyntax)
+    }
+    srd.schemaDefinitionUnless(vrd.primType == PrimType.UnsignedInt,
+      "Variable '%s' is not of type 'xs:unsignedInt'.", varQName.toExtendedSyntax)
+    vrd
+  }
+
+  private var limitingOutputStream : ByteBufferOutputStream = _
+  private var originalOutputStream: OutputStream = _
+
+  protected def wrapLayerDecoder(jis: InputStream) = jis
+
+  override def wrapLimitingStream(jis: java.io.InputStream, state: PState) = {
+    val layerLengthInBytes: Int = layerLengthEv.evaluate(state).toInt
+    val ba = new Array[Byte](layerLengthInBytes)
+    try
+      IOUtils.readFully(jis, ba)
+    catch {
+      case eof: EOFException =>
+        throw DataInputStream.NotEnoughDataException(ba.length * 8)
+    }
+    // this stream will be used by the parse when it parses the
+    // schema contents of the layer. We have already consumed the bytes
+    // from the original input stream, and saved those in the byte array
+    // to enable the computation of the checkDigit.
+    val layerStream = new ByteArrayInputStream(ba)
+    val str = new String(ba, charset) // always does replacement on decode errors.
+    val checkDigit = computeCheckDigit(str)
+    state.setVariable(checkDigitVRD, checkDigit, srd) // assign to result variable
+    layerStream
```

Comment:
Is this standard for layers, to read the entire data into memory, do something with it, and then return it as a new stream that parse then reads? Or is that just done for simplicity because this is a test implementation so not a big deal, or we expect the size of these layers to be small and not be a memory issue?

Reply:
If the checksum was over a large thing, like a whole file, then one would want to use a layered stream that does the checksumming as part of the parser reading the data stream. That would avoid the need to buffer stuff. That's what for example the gzip and line-folding layers do. 

In this case this is IP packets, by definition small. So just grabbing it into a buffer seemed simplest.